python/ray/spark/utils.py

```diff
@@ -0,0 +1,261 @@
+
+    stdout, stderr = process.communicate()
+    returncode = process.poll()
+    comp_process = subprocess.CompletedProcess(
+        process.args,
+        returncode=returncode,
+        stdout=stdout,
+        stderr=stderr,
+    )
+    if throw_on_error and returncode != 0:
+        raise ShellCommandException.from_completed_process(comp_process)
+    return comp_process
+
+
+def get_safe_port():
+    """Returns an ephemeral port that is very likely to be free to bind to."""
+    import socket
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        sock.bind(("0.0.0.0", 0))
+        return sock.getsockname()[1]
+
+
+def get_random_port(min_port=20000, max_port=60000):
+    rng = random.SystemRandom()
+    return rng.randint(min_port, max_port)
+
+
+def check_port_open(host, port):
+    import socket
+    from contextlib import closing
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        return sock.connect_ex((host, port)) == 0
+
+
+def get_spark_session():
+    from pyspark.sql import SparkSession
+    return SparkSession.builder.getOrCreate()
+
+
+def get_spark_driver_hostname(spark_master_url):
+    if spark_master_url.lower().startswith("local"):
+        return "127.0.0.1"
+    else:
+        parsed_spark_master_url = urlparse(spark_master_url)
+        if parsed_spark_master_url.scheme.lower() != "spark" or \
+                not parsed_spark_master_url.hostname:
+            raise ValueError(f"Unsupported spark.master URL: {spark_master_url}")
+        return parsed_spark_master_url.hostname
+
+
+def get_max_num_concurrent_tasks(spark_context):
+    """Gets the current max number of concurrent tasks."""
+    # pylint: disable=protected-access
+    # spark 3.1 and above has a different API for fetching max concurrent tasks
+    if spark_context._jsc.sc().version() >= "3.1":
+        return spark_context._jsc.sc().maxNumConcurrentTasks(
+            spark_context._jsc.sc().resourceProfileManager().resourceProfileFromId(0)
+        )
+    return spark_context._jsc.sc().maxNumConcurrentTasks()
+
+
+def _get_total_phyisical_memory():
+    import psutil
+    return psutil.virtual_memory().total
+
+
+def _get_total_shared_memory():
+    import shutil
+    return shutil.disk_usage('/dev/shm').total
+
+
+def _get_cpu_cores():
+    import multiprocessing
+    return multiprocessing.cpu_count()
+
+
+def _calc_mem_per_ray_worker(num_task_slots, physical_mem_bytes, shared_mem_bytes):
+    ray_worker_object_store_bytes = int(shared_mem_bytes / num_task_slots * 0.8)
+    ray_worker_heap_mem_bytes = int((physical_mem_bytes - shared_mem_bytes) / num_task_slots * 0.8)
+    return ray_worker_heap_mem_bytes, ray_worker_object_store_bytes
+
+
+def get_avail_mem_per_ray_worker(spark):
+    """
+    Return the available heap memory and object store memory for each ray worker.
+    NB: We have one ray node per spark task.
+    """
+    # TODO: add a option of heap memory / object store memory ratio.
```

Comment:
On databricks runtime, the shared memory (/dev/shm) capacity is about 50% ~ 60% of the total physical memory.
I think the ratio is too high, so I suggest to add a option to control the ratio.

Reply:
Added arg `heap_to_object_store_memory_ratio` in `init_cluster()` to control this and used your above logic to handle the ratio allowance configured for heap vs. object store memory.